File: ApiLib.py

#############################
# 0day.today unofficial api #
#       By MrSentex         #
#############################

# <= Imports =>
from random import SystemRandom
from requests import Session
from codecs import ascii_encode
from unidecode import unidecode
from bs4 import BeautifulSoup as Soup
# <= Imports =>

class api_0day_today:

    def __init__(self):

        self.url = "http://{}.0day.today".format(self.randomSubDomain(5))
        self.session_obj = Session()

        self.terms = self.acceptTerms()
    
    def acceptTerms(self):
        try:
            self.session_obj.post(self.url, data={"agree" : "Yes, I agree"})
            return (True, None)
        except Exception as err:
            return (False, str(err))

    def randomSubDomain(self, interactions):

        string = ""

        for _ in range(1, interactions):
            string += SystemRandom().choice("abcdefghijklmnopqrstuvwxyz1234567890")
        
        return string
    
    def fixString(self, string):

        string = ascii_encode(unidecode(string.replace("\n", "").replace("\t", "")))[0]

        if string.find("Comments:") != -1:
            string = string[:string.find("Comments: {}".format(string[string.find("Comments:")+10:string.find("Comments:")+len(string)-string.find("Comments:")-1]))]

        if string.find("Rate down:") != -1:
            string = string[:string.find("Rate down: {}".format(string[string.find("Rate down:")+11:string.find("Rate down:")+len(string)-string.find("Rate down:")-1]))]
        
        if string.find("Rate up:") != -1:
            string = string[:string.find("Rate up: {}".format(string[string.find("Rate up:")+9:string.find("Rate up:")+len(string)-string.find("Rate up:")-1]))]

        return string
    
    def fixPrice(self, price):

        price = self.fixString(price)

        if price.startswith("free"):
            return "free"

        btc = price[price.find("for")+4:price.find("BTC")-1]
        pre_gold = price[price.find("BTC")+4:len(price)]
        gold = pre_gold[pre_gold.find("for")+4:pre_gold.find("GOLD")-1]
        return "{} BTC or {} GOLD".format(btc, gold)

    def search(self, param):

        if not self.terms[0]:
            return {"status" : "fail", "exception" : "An error ocurred in the acceptTerms function | {}".format(self.terms[1])}

        param = param.replace(" ", "+")

        try:

            response = self.session_obj.get("{}/search?search_request={}".format(self.url, param))
            parser = Soup(response.text, "lxml")

            result = []

            tables = parser.find_all("div", attrs={"class" : "ExploitTableContent"})

            for table in tables:
                rows = table.find_all("div", attrs={"class" : "td"})

                date = self.fixString(rows[0].getText())
                desc = self.fixString(rows[1].getText())
                platform = self.fixString(rows[2].getText())
                price = self.fixPrice(rows[9].getText())
                author = self.fixString(rows[10].getText())[0:self.fixString(rows[10].getText()).find("Exploits")]
                url = rows[1].find("a", href=True)["href"].replace("/description", "")

                result.append({"date" : date, "desc" : desc, "platform" : platform, "price" : price, "author" : author, "url" : self.url+url})
        
            return {"status" : "success", "response" : result}

        except Exception as err:
            return {"status" : "fail", "exception" : str(err)}
    
    def getIndex(self):

        if not self.terms[0]:
            return {"status" : "fail", "exception" : "An error ocurred in the acceptTerms function | {}".format(self.terms[1])}

        try:

            response = self.session_obj.get("{}".format(self.url))
            parser = Soup(response.text, "lxml")

            result = []

            tables = parser.find_all("div", attrs={"class" : "ExploitTableContent"})

            for table in tables:
                rows = table.find_all("div", attrs={"class" : "td"})

                date = self.fixString(rows[0].getText())
                desc = self.fixString(rows[1].getText())
                platform = self.fixString(rows[2].getText())
                price = self.fixPrice(rows[9].getText())
                author = self.fixString(rows[10].getText())[0:self.fixString(rows[10].getText()).find("Exploits")]
                url = rows[1].find("a", href=True)["href"].replace("/description", "")

                result.append({"date" : date, "desc" : desc, "platform" : platform, "price" : price, "author" : author, "url" : self.url+url})
        
            return {"status" : "success", "response" : result}

        except Exception as err:
            return {"status" : "fail", "exception" : str(err)}

File: ApiLib.php
<?php

    include("parser.php");

    class api_0day_today {

        function __construct() {

            $this->url = sprintf("http://%s.0day.today/", 8, $this->randomSubDomain(5));

            $this->curl_obj = curl_init();
            curl_setopt($this->curl_obj, CURLOPT_COOKIEJAR, dirname(__FILE__) . '/cookie.txt');
            curl_setopt($this->curl_obj, CURLOPT_RETURNTRANSFER,1);

            $this->terms = $this->acceptTerms();

            $this->platforms = array("aix", "Android", "bsd", "freebsd", "hp-ux", "iOS", "irix", "linux", "macOS", "minix", "netware", "novell", "openbsd", "plan9", "QNX", "sco", "solaris", "Symbian", "tru64", "ultrix", "unix", "windows", "hardware", "multiple", "unsorted", "aix", "alpha", "arm", "bsd", "bsd/ppc", "bsd/x86", "bsdi/x86", "freebsd/x86", "freebsd/x86-64", "generator", "hardware", "hpux", "irix", "linux/amd64", "linux/mips", "linux/ppc", "linux/sparc", "linux/x86", "linux/x86-64", "multiple", "netbsd/x86", "openbsd/x86", "os-x/ppc", "os-x/x86", "sco/x86", "solaris/sparc", "solaris/x86", "unixware", "win32", "win64", "asp", "cgi", "java", "jsp", "perl", "php", "python", "ruby", "tricks", "xml");

        }

        function acceptTerms() {
            try {

                curl_setopt($this->curl_obj, CURLOPT_URL, $this->url);
                curl_setopt($this->curl_obj, CURLOPT_POST, true);
                curl_setopt($this->curl_obj, CURLOPT_POSTFIELDS, "agree=Yes, I agree");
                $ot = curl_exec($this->curl_obj);

                curl_setopt($this->curl_obj, CURLOPT_POST, false);
                
                $s_code = curl_getinfo($this->curl_obj, CURLINFO_HTTP_CODE);

                if ($s_code != 200) {
                    return array(false, "Status code: ".$s_code);
                }

                return array(true, null);

            } catch (Exception $err) {
                return array(false, $err->getMessage());
            }

        }

        function randomSubDomain($interactions) {

            $string = "";

            foreach(range(1, $interactions) as $n) {
                $string .= "abcdefghijklmnopqrsuvwxzy1234567890"[rand(0, 35)];
            }

            return $string;

        }

        function findUrl($array) {
            foreach($array as $hit) {

                if (strpos($hit->href, "exploit")) {
                    $url = explode("/", $hit->href);
                    return $url[count($url)-1];
                }

            }
        }

        function fixArray($array_str, $url) {

            $array = explode(" ", $array_str);
            $array_fix = array();

            for($i = 0; $i < count($array); $i++) {
                if (trim($array[$i]) !== "") {
                    $array_fix[$i] = trim($array[$i]);
                }
            }

            $array = $array_fix;

            $date = $array[0];
            $desc = "";

            for($i = 1; $i < count($array); $i++) {
                if ($array[$i] === "Comments:") {
                    $last = $i;
                    break;
                }

                $desc .= $array[$i]." ";

            }

            for($i = $last; $i < count($array); $i++) {
                if (in_array($array[$i], $this->platforms)) {
                    $platform = $array[$i];
                    $last = $i;
                    break;
                }
            }

            if (!in_array("BTC", $array)) {
                $price = "free";
            } else {

                for($i = $last; $i < count($array); $i++) {

                    if ($array[$i] === "BTC") {
                        $btc = $array[$i-1];
                    }

                    if ($array[$i] === "GOLD") {
                        $gold = $array[$i-1];
                    }

                    if(isset($btc) && isset($gold)) {
                        $last = $i;
                        $price = $btc." BTC or ".$gold." GOLD";
                        break;
                    }

                }

            }

            $name = "";

            for($i = $last; $i < count($array); $i++) {

                if($array[$i] === "Exploits:") {
                    $l = $i;
                    for($i = $l-3; $i < $l; $i++) {

                        if ($array[$i] !== "free" && $array[$i] !== "GOLD") {
                            $name .= $array[$i]." ";
                        }

                    }
                }

            }

            $name = utf8_decode(substr($name, 0, count($name)-2));
            $desc = substr($desc, 0, count($desc)-2);

            return array(
                "date" => $date,
                "desc" => $desc,
                "platform" => $platform,
                "price" => $price,
                "author" => $name,
                "url" => "https://0day.today/exploit/".$url
            );

        }

        function search($param) {

            if (!$this->terms[0]) {
                return array(
                    "status" => "fail",
                    "exception" => "An error ocurred in the acceptTerms function | ".$this->terms[1]
                );
            }

            $param = str_replace(" ", "+", $param);

            try {

                curl_setopt($this->curl_obj, CURLOPT_URL, $this->url."search?search_request=".$param);
                $response = curl_exec($this->curl_obj);

                $exploits = array();

                if (!$response) {
                    return array(
                        "status" => "fail",
                        "exception" => "Bad response form 0day.today"
                    );
                }

                $parser = str_get_html($response);

                foreach($parser->find("div") as $divs) {
                    if ($divs->class === "ExploitTableContent") {
                        $url = $this->findUrl($divs->find("a"));
                        $tmp_parser = str_get_html($divs->plaintext);
                        $data = $tmp_parser->plaintext;
                        array_push($exploits, $this->fixArray($data, $url));
                    }
                }


                return array(
                    "status" => "success",
                    "response" => $exploits
                );

            } catch (Exception $err) {
                return array(
                    "status" => "fail",
                    "exception" => "Scrape error | ".$err->getMessage()
                );
            }

        } 
    }
?>